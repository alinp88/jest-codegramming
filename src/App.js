import { useState } from "react";
import "./App.css";
import styled from "styled-components";
import { passwordStrength, validateUser } from "./utils.ts";

const LoginHolder = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`;

const LoginBox = styled.div`
  width: 400px;
  height: 400px;
  text-align: left;

  span {
    display: inline-block;
    width: 100px;
    margin-right: 25px;
  }

  input {
    width: 250px;
  }

  .good {
    color: green;
  }

  .bad {
    color: red;
  }
`;

function App() {
  let [username, setUsername] = useState('');
  let [password, setPassword] = useState('');
  let [loginState, setLoginState] = useState('');

  return (
    <LoginHolder>
      <LoginBox>
        <h1>App Login</h1>
        <hr />
        <span>username</span>
        <input
          type="text"
          aria-label="username"
          value={username}
          onChange={(event) => setUsername(event.target.value)}
        ></input>
        <br />
        <span>password</span>
        <input
          type="password"
          aria-label="password"
          value={password}
          onChange={(event) => setPassword(event.target.value)}
        ></input>
        <br />
        <span>strength</span>
        <span>{password && passwordStrength(password)}</span>
        <br/>
        <hr/>
        <button onClick={() => {
          setLoginState(validateUser(username, password) ? 'good' : 'bad');
        }}>Login</button>
        <br/>
        {loginState && <p className={loginState}>
          login state: {loginState}
        </p>}
      </LoginBox>
    </LoginHolder>
  );
}

export default App;
