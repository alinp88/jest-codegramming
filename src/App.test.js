/**
 * @group ui
 * @group all
 */

import { render, screen, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import App from './App';

describe ('Main App', () => {
  describe('Login Screen', () => {
    test('Title', () => {
      render(<App />);
      const element = screen.getByText(/App Login/i);
      expect(element).toBeInTheDocument();
    });
    test('Inputs', () => {
      render(<App />);
      const element1 = screen.getByText(/username/i);
      expect(element1).toBeInTheDocument();
      const element2 = screen.getByText(/password/i);
      expect(element2).toBeInTheDocument();
    });
    test('Password Strength', () => {
      render(<App />);
      const element = screen.getByText(/strength/i);
      expect(element).toBeInTheDocument();
    });
  })
});

describe ('Login Test', () => {
  test('Login Success', () => {
    const app = render(<App />);
    const username = app.getByLabelText('username');
    fireEvent.change(username, {target: {value: 'good'}})
    const password = app.getByLabelText('password');
    fireEvent.change(password, {target: {value: 'good_password'}})
    const loginButton = app.getByText('Login');
    fireEvent.click(loginButton);
    const element = screen.getByText(/login state: good/i);
    expect(element).toBeInTheDocument();
    expect(element).toHaveClass('good');
  });
  test('Login Fail', () => {
    const app = render(<App />);
    const username = app.getByLabelText('username');
    fireEvent.change(username, {target: {value: 'bad'}})
    const password = app.getByLabelText('password');
    fireEvent.change(password, {target: {value: 'bad_password'}})
    const loginButton = app.getByText('Login');
    fireEvent.click(loginButton);
    const element = screen.getByText(/login state: bad/i);
    expect(element).toBeInTheDocument();
    expect(element).toHaveClass('bad');
  });
});

test('Mocked function call', () => {
  const mock = jest.fn()
  mock(1);
  mock(2);
  expect(mock).toBeCalledWith(1);
  expect(mock).toBeCalledTimes(2);
});

test('App renders correctly', () => {
  const tree = renderer
    .create(<App/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
  //expect(tree).toMatchInlineSnapshot();
});
