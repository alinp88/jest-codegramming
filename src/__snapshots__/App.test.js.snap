// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  width: 400px;
  height: 400px;
  text-align: left;
}

.c1 span {
  display: inline-block;
  width: 100px;
  margin-right: 25px;
}

.c1 input {
  width: 250px;
}

.c1 .good {
  color: green;
}

.c1 .bad {
  color: red;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <h1>
      App Login
    </h1>
    <hr />
    <span>
      username
    </span>
    <input
      aria-label="username"
      onChange={[Function]}
      type="text"
      value=""
    />
    <br />
    <span>
      password
    </span>
    <input
      aria-label="password"
      onChange={[Function]}
      type="password"
      value=""
    />
    <br />
    <span>
      strength
    </span>
    <span />
    <br />
    <hr />
    <button
      onClick={[Function]}
    >
      Login
    </button>
    <br />
  </div>
</div>
`;
